pkg/review/job.go

```diff
@@ -81,61 +75,78 @@ func ExecuteJob(ctx context.Context, cfg *Config) error {
 	}
 
 	// Step 2: Get review status information for each commit.
-	// Fan out the work of processing all of the commits that were found
-	for _, commit := range commits {
-		if err := commitPool.Do(ctx, func() (*CommitReviewStatus, error) {
-			result := processCommit(ctx, *commit, gitHubClient)
-			return result, nil
-		}); err != nil {
-			return fmt.Errorf("failed to submit job to worker pool: %w", err)
-		}
-	}
-	// When all of the workers are complete, extract the result values
-	results, err := commitPool.Done(ctx)
+	commitReviewStatuses, err := pooledTransform(ctx, commits,
+		func(commit *Commit) (*CommitReviewStatus, error) {
+			return processCommit(ctx, gitHubClient, commit), nil
+		},
+	)
 	if err != nil {
 		return fmt.Errorf("failed to process commits: %w", err)
 	}
-	statuses := make([]*CommitReviewStatus, 0, len(results))
-	for _, v := range results {
-		// Statuses that were failed to be retrieved will be nil
-		if v.Value != nil {
-			statuses = append(statuses, v.Value)
-		}
+	// statuses that should not be further processed will be nil, so we should exclude them
+	commitReviewStatuses = removeNil(commitReviewStatuses)
+
+	// Step 3: Look up break glass issue if necessary and tag the review status with it if found.
+	fetcher := &BigQueryBreakGlassIssueFetcher{
+		client: bqClient,
+	}
+	taggedReviewStatuses, err := pooledTransform(ctx, commitReviewStatuses,
+		func(status *CommitReviewStatus) (*CommitReviewStatus, error) {
+			return processReviewStatus(ctx, fetcher, cfg, status), nil
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to process commit review statuses: %w", err)
+	}
+	// statuses that should not be further processed will be nil
+	taggedReviewStatuses = removeNil(taggedReviewStatuses)
```

Comment:
Why would any of these return nil? If they do, it'd be better to skip them on the next iteration instead of making an entire copy of the slice with nils removed.

Reply:
This is due to the underly processing functions [processCommit](https://github.com/abcxyz/github-metrics-aggregator/blob/main/pkg/review/commit_review_status.go#L146) and [processReviewStatus](https://github.com/abcxyz/github-metrics-aggregator/blob/main/pkg/review/commit_review_status.go#L240) both using `nil` as signal that the item should not be further processed.

I agree that copying the entire slice is not the best approach. I will either make a new PR to address this or include a fix for this in the resubmitted PR if we decide to revert this one.